### PR TITLE
Preliminary constraint solver support for variadic generics

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/AST/ParameterPack.cpp [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    ++j;
+  }
+
+  auto *packType = PackType::get(getASTContext(), expandedTypes);
+  return PackExpansionType::get(packType, countType);
+}
+
+bool TupleType::containsPackExpansionType() const {
+  for (auto elt : getElements()) {
+    if (elt.getType()->is<PackExpansionType>())
```
Comment (xedin) [inline lib/AST/ParameterPack.cpp:114]: Why not make this a property of a tuple type?

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -6898,6 +7034,20 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     }
   }
 
+  if (isa<PackExpansionType>(desugar1) && isa<PackType>(desugar2)) {
+    auto *packExpansionType = cast<PackExpansionType>(desugar1);
+    auto *packType = cast<PackType>(desugar2);
+
+    if (packExpansionType->getPatternType()->is<TypeVariableType>())
+      return matchTypes(packExpansionType->getPatternType(), packType, kind, subflags, locator);
```
[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/Sema/CSSimplify.cpp:7042]: It was a hack to work around other problems in PackExpansionMatcher.cpp. This chunk of code was removed in subsequent PRs.

File: lib/AST/ParameterPack.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/AST/ParameterPack.cpp:114]: That's a good idea. I wanted to wait to micro-optimize until more code is fleshed out though.

**Pull request merged**